File: app/src/main/java/com/porarri/yamabikochat/data/AutoConversationFormatter.kt
package com.porarri.yamabikochat.data

import com.porarri.yamabikochat.data.local.AutoConversationMessage

/**
 * Formats auto conversation content for display while keeping the raw text that should be fed back
 * into language models free from hidden reasoning traces.
 */
fun formatAutoConversationDisplay(content: String, reasoning: String?): String {
    val trimmedContent = content.trim()
    val trimmedReasoning = reasoning?.trim().orEmpty()

    return when {
        trimmedContent.isEmpty() && trimmedReasoning.isEmpty() -> ""
        trimmedReasoning.isEmpty() -> trimmedContent
        trimmedContent.isEmpty() -> buildThinkingBlock(trimmedReasoning)
        else -> buildString {
            append(trimmedContent)
            append("\n\n")
            append(buildThinkingBlock(trimmedReasoning))
        }
    }
}

fun AutoConversationMessage.displayContent(): String =
    formatAutoConversationDisplay(content, reasoning)

private fun buildThinkingBlock(reasoning: String): String {
    return buildString {
        append("```thinking\n")
        append(reasoning)
        append("\n```")
    }
}

File: ios/YamabikoChat/Data/Models/AutoConversationFormatter.swift
import Foundation

func formatAutoConversationDisplay(content: String, reasoning: String?) -> String {
    let trimmedContent = content.trimmingCharacters(in: .whitespacesAndNewlines)
    let trimmedReasoning = (reasoning ?? "").trimmingCharacters(in: .whitespacesAndNewlines)

    switch (trimmedContent.isEmpty, trimmedReasoning.isEmpty) {
    case (true, true):
        return ""
    case (false, true):
        return trimmedContent
    case (true, false):
        return buildThinkingBlock(trimmedReasoning)
    case (false, false):
        return "\(trimmedContent)\n\n\(buildThinkingBlock(trimmedReasoning))"
    }
}

private func buildThinkingBlock(_ reasoning: String) -> String {
    "```thinking\n\(reasoning)\n```"
}
